prov/sm2/src/sm2_util.c

```diff
@@ -0,0 +1,169 @@
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+
+struct dlist_entry sm2_ep_name_list;
+DEFINE_LIST(sm2_ep_name_list);
+pthread_mutex_t sm2_ep_list_lock = PTHREAD_MUTEX_INITIALIZER;
+
+void sm2_cleanup(void)
+{
+	struct sm2_ep_name *ep_name;
+	struct dlist_entry *tmp;
+
+	pthread_mutex_lock(&sm2_ep_list_lock);
+	dlist_foreach_container_safe (&sm2_ep_name_list, struct sm2_ep_name,
+				      ep_name, entry, tmp)
+		free(ep_name);
+	pthread_mutex_unlock(&sm2_ep_list_lock);
+}
+
+size_t sm2_calculate_size_offsets(ptrdiff_t *rq_offset, ptrdiff_t *fq_offset)
+{
+	size_t total_size;
+
+	/* First memory block.  The header of an sm2_region */
+	total_size = sizeof(struct sm2_region);
+
+	/* Second memory block: the recv_queue_fifo, an sm2_fifo */
+	if (rq_offset)
+		*rq_offset = total_size;
+	total_size += sizeof(struct sm2_fifo);
+
+	/* Third memory block: the message objects in a free queue */
+	if (fq_offset)
+		*fq_offset = total_size;
+	total_size += freestack_size(sizeof(struct sm2_xfer_entry),
+				     SM2_NUM_XFER_ENTRY_PER_PEER);
+
+	/*
+	 * Revisit later to see if we really need the size adjustment, or
+	 * at most align to a multiple of a page size.
+	 */
+	total_size = roundup_power_of_two(total_size);
+
+	return total_size;
+}
+
+int sm2_create(const struct fi_provider *prov, const struct sm2_attr *attr,
+	       struct sm2_mmap *sm2_mmap, int *gid)
+{
+	struct sm2_ep_name *ep_name;
+	ptrdiff_t recv_queue_offset, free_stack_offset;
+	int ret;
+	void *mapped_addr;
+	struct sm2_region *smr;
+
+	sm2_calculate_size_offsets(&recv_queue_offset, &free_stack_offset);
+
+	FI_WARN(prov, FI_LOG_EP_CTRL, "Claiming an entry for (%s)\n",
+		attr->name);
+	sm2_file_lock(sm2_mmap);
+	ret = sm2_entry_allocate(attr->name, sm2_mmap, gid, true);
+
+	/* TODO: handle address-in-use error (FI_EBUSY?)*/
+	/* TODO: handle no available space on device error */
+	/* TODO: handle no available slots left error */
+
+	ep_name = calloc(1, sizeof(*ep_name));
+	if (!ep_name) {
+		FI_WARN(prov, FI_LOG_EP_CTRL, "calloc error\n");
+		ret = -FI_ENOMEM;
+		goto close;
+	}
+	strncpy(ep_name->name, (char *) attr->name, SM2_NAME_MAX - 1);
+	ep_name->name[SM2_NAME_MAX - 1] = '\0';
+
+	pthread_mutex_lock(&sm2_ep_list_lock);
+	dlist_insert_tail(&ep_name->entry, &sm2_ep_name_list);
+
+	if (ret < 0) {
+		FI_WARN(prov, FI_LOG_EP_CTRL, "ftruncate error\n");
+		ret = -errno;
+		goto remove;
+	}
+
+	mapped_addr = sm2_mmap_ep_region(sm2_mmap, *gid);
+
+	if (mapped_addr == MAP_FAILED) {
+		FI_WARN(prov, FI_LOG_EP_CTRL, "mmap error\n");
+		ret = -errno;
+		goto remove;
+	}
+
+	/* TODO: Handle SM2_FLAG_HMEM_ENABLED */
+
+	pthread_mutex_unlock(&sm2_ep_list_lock);
+	smr = mapped_addr;
+
+	smr->version = SM2_VERSION;
+	smr->flags = attr->flags;
+	smr->recv_queue_offset = recv_queue_offset;
+	smr->free_stack_offset = free_stack_offset;
+
+	sm2_fifo_init(sm2_recv_queue(smr));
+	smr_freestack_init(sm2_free_stack(smr), SM2_NUM_XFER_ENTRY_PER_PEER,
```

Comment:
Since the freestack is only accessed by the owner, you can use the regular freestack implementation instead of the shm-supported one (which has extra checks and pointers so multiple processes can access it). see OFI_DECLARE_FREESTACK in include/ofi_mem

Reply:
oooh this might speed us up!

Reply:
Request to do this as a follow on PR b/c this isn't trivial.